hpct-ldap-server/src/servers/ldap.py

```diff
@@ -0,0 +1,503 @@
+
+from charms.operator_libs_linux.v0 import apt
+from charms.operator_libs_linux.v1 import systemd
+
+from utils.filedata import FileData
+
+MIN_GID = 999
+MIN_UID = 999
+
+
+class LdapServer:
+    """Server to provide LDAP server charm all functionality needed."""
+
+    packages = ["slapd", "ldap-utils", "sssd-ldap", "gnutls-bin", "ssl-cert"]
+    systemd_services = ["slapd"]
+
+    def __init__(self):
+        pass
+
+    def _add_base(self, admin_passwd, dcs) -> None:
+        """Define base for groups and users.
+
+        Parameters
+        ----------
+        admin_passwd : str
+            LDAP password.
+        dcs : str
+            Domain components.
+        """
+        base = [
+            f"dn: ou=People,{dcs}",
+            "objectClass: organizationalUnit",
+            "ou: People",
+            "",
+            f"dn: ou=Groups,{dcs}",
+            "objectClass: organizationalUnit",
+            "ou: Groups",
+        ]
+
+        with open("/etc/ldap/basedn.ldif", "w") as f:
+            for line in base:
+                f.write(f"{line}\n")
+
+        binddn = f"cn=admin,{dcs}"
+        cmd = ["ldapadd", "-x", "-D", binddn, "-w", admin_passwd, "-f", "/etc/ldap/basedn.ldif"]
+
+        rc = subprocess.call(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.STDOUT)
+        if rc != 0:
+            raise Exception("Unable to add basedn ldif.")
+
+    def _split_domain(self, dc) -> str:
+        """Split domain components.
+
+        Parameters
+        ----------
+        dcs : str
+            Domain components.
+        """
+        if "." in dc:
+            split_dc = []
+            dcs = dc.split(".")
+            for d in dcs:
+                if len(d) < 1:
+                    raise Exception("Invalid domain components given, cannot be empty.")
+                else:
+                    split_dc.append(f"dc={d}")
+            return ",".join(split_dc)
+        else:
+            return f"dc={dc}"
+
+    def add_user(
+        self,
+        admin_passwd=None,
+        domain=None,
+        gecos=None,
+        gid=None,
+        homedir=None,
+        login=None,
+        passwd=None,
+        uid=None,
+        user=None,
+    ) -> None:
+        """Add user.
+
+        Parameters
+        ----------
+        admin_passwd : str
+            LDAP password.
+        domain : str
+            Domain name.
+        gecos : str
+            Real name.
+        gid : int
+            Group id.
+        homedir : str
+            Home directory path.
+        login : str
+            Login shell.
+        passwd : str
+            User password.
+        uid : int
+            User id.
+        user : str
+            Username.
+        """
+        dcs = self._split_domain(domain)
+
+        if (
+            None not in [admin_passwd, dcs, gecos, gid, homedir, login, passwd, uid, user]
+            and uid > MIN_UID
+            and gid > MIN_GID
+        ):
+            binddn = f"cn=admin,{dcs}"
```

Comment:
You could check for unacceptable values and return immediately (with error?). That would allow you to dedent the following.

Reply:
Updated as suggested